### PR TITLE
Schedule dynamic filtering collecting task immediately

### PR DESCRIPTION
**Pull request by sopel39**
In case of plan
        J1
       /  \
     J2    S3
    /  \
   S1  S2

It might happen that that dynamic filtering dependencies are:
S3 => S3 => S1

With phased scheduler source stage consisting of
(J1, J2, S1) won't be scheduler until stages with S3 and S2
have finished split enumaration. However, it might happen
that S2 is waiting for dynamic filters for S3. In that case,
S3 will never complete because DFs for S3 are collected in
stage (J1, J2, S1).

This commit makes scheduling of DF collecting task immediately
which will prevent queries from deadlock.

### CODE REVIEW

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java
Comment (raunaqmorarka) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java:78]: It would be good to have a comment here that the expression `(k0 * -1) + 2 = i2` is being used instead of `k0 = i2` to avoid `t0` table scan having dependency on a DF

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java
Comment (sopel39) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java]: I've added `createPartitionedTable` abstract method.

**Review by raunaqmorarka — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergDynamicPartitionPruningTest.java
Comment (dain) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergDynamicPartitionPruningTest.java:42]: There is a typo in the commit message

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:549]: In the old code this would return `Optional<RemoteTask>`.  Do we not need to handle the case where we got `Optional.empty()`?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:244]: The old code as adding this task to `overallNewTasks`.  Do we not need to track this anymore?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:241]: I suggest we are more direct with this statement.  Something like
```
Avoid deadlocks by immediately scheduling a task for collecting dynamic filters, because there can be task in other stages blocked waiting for the dynamic filters.
```

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:243]: If this is a new restriction, make sure to document it.

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:242]: I don't understand this change at all.  Someone else should review it.

**Review by dain — COMMENTED** [review]
The scheduler stuff and the tests look good to me.  I don't understand the implications of the change to `DynamicFilterService`.

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (joshthoward) [reply core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:242]: @raunaqmorarka can you describe this in more detail so that we can validate that this change is okay tomorrow?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:245]: no, `io.trino.execution.scheduler.ScheduleResult#getNewTasks` is only used in tests.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:244]: It's not new. We have one in `io/trino/sql/planner/LocalExecutionPlanner.java:2871` already

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:550]: In old code this was used to return new tasks in `io.trino.execution.scheduler.ScheduleResult`. However, `io.trino.execution.scheduler.ScheduleResult#getNewTasks` is only used for testing purposes

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java]: The change essentially does: If this is a source stage and some dynamic filters is produced by this stage (for broadcast join), then we need extra DF collecting task.

I've updated comment here:
```
            // Dynamic filters can be collected by single task only for source distribution (non-fixed) stage.
            // Such stage must only contain replicated joins.
```

**Comment by sopel39** [comment]
ac @raunaqmorarka @dain . I've narrowed down when phased scheduler won't start join stage immediately to only non-fixed source stages. I've also improved tests with regards to source stage partitioning

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (dain) [inline core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:499]: you could inline this

**Review by dain — APPROVED** [review]
looks good

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:244]: I'm not sure if scheduling is supposed to happen here. Could you please explain why is it necessary to schedule a task in `startLifespan`? Isn't the `schedule` method going to be called at least once before blocking?

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:505]: I assume this change is just a refactor. You just added an assertion that `producedDynamicFilters` are equal to `replicatedDynamicFilters`. And given `producedDynamicFilters` and `replicatedDynamicFilters` sets are equal, `producedDynamicFilters` x `consumedDymanicFilters` x `replicatedDynamicFilters` = `producedDynamicFilters` x `consumedDymanicFilters`.

Is my understanding correct?

File: core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:241]: This changes the logic from `If there's any dynamic filters produced and consumed within a stage` to `If there's any dynamic filters produced by a stage`. Am i correct? Could you please explain why is it necessary? Is it needed to cover a mix of broadcast and partitioned joins?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:415]: From what I understand it is to prevent a non source distributed stage running a replicated join from being executed lazily. Is it correct? I wonder why is it necessary.

**Review by arhimondr — COMMENTED** [review]
It doesn't feel like I really understand the problem. Added some questions. 

It would be great if you could extract refactoring and improvements that are unrelated to have a commit that only does what's needed to address the problem. 

Additionally if you could elaborate more on a problem in the commit message it would be great. Maybe you can provide an example of a query that triggers a deadlock with a description of what join distributions are, where are the stage boundaries and how dynamic filters interact with each other.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:244]: It's needed to collect dynamic filters for source stages with broadcast joins. 

> Isn't the schedule method going to be called at least once before blocking?

No. With phased scheduler some stages won't be scheduled before all dependent stages enumerate all splits. In this case there might be deadlock (without this extra task), because DF is collected by donwstream stage.

**Comment by sopel39** [comment]
@arhimondr I've simplified this PR (removed refactor), keeping just needed parts. i cannot answer some outdated comments

> This changes the logic from If there's any dynamic filters produced and consumed within a stage to If there's any dynamic filters produced by a stage. Am i correct? Could you please explain why is it necessary? Is it needed to cover a mix of broadcast and partitioned joins?

I've changed to logic to create collecting task if there is any lazy DF produced by stage. This is needed because there might be consumers of that lazy DF outside of stage.

> From what I understand it is to prevent a non source distributed stage running a replicated join from being executed lazily. Is it correct? I wonder why is it necessary.

partitioned stages wont be covered by extra collecting task. Hence, if there are any lazy DFs produced by that stage, then stage needs to be scheduled

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:244]: This will effectively schedule a task on `StageScheduler` construction before `SqlQueryScheduler` is fully initialized and scheduling is started. 

At very least the state reporting will be inconsistent. The fact that there are some tasks running while the stage is still in a state `PLANNED` might be very non obvious and potentially error prone. Also it's not clear what is going to happen if a scheduled task fails or finishes before `SqlQueryScheduler` is fully initialized and all the listeners are attached. Even if it works today, it might easily break in the future, as it might not be expected for any tasks to exist before the `SqlQueryScheduler` is initialized and the stage is in state `SCHEDULING`. I would strongly recommend to avoid this route.

It generally feels like it should be responsibility of the `ExecutionSchedule` (`PhasedExecutionSchedule`) to determine what stages should be scheduled in what order. I was going to recommend making the `PhasedExecutionSchedule` aware of dynamic filters but it seems like you are already doing it. Based on the dynamic filter dependencies the `PhasedExecutionSchedule` can trigger scheduling for a source stage that contains broadcast join. Then the `SourcePartitionedScheduler` can rely on the existing mechanism on scheduling a single task when the split enumeration is blocked.

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java]: I've added `io.trino.execution.scheduler.SourceScheduler#start` which is called from within `io.trino.execution.scheduler.SqlQueryScheduler.PipelinedDistributedStagesScheduler#schedule` for every `SourceScheduler`.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1545]: nit: maybe move it inside the `try` block

File: core/trino-main/src/main/java/io/trino/execution/scheduler/StageScheduler.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/StageScheduler.java:21]: Maybe add a short documentation, e.g.:

`Called by the query scheduler when the scheduling process begins. This method is called before the ExecutionSchedule takes a decision to schedule a stage but after the query scheduling has been fully initialized. Within this method the scheduler may decide to schedule tasks that are necessary for query execution to make progress. For example the the scheduler may decided to schedule a task without assigning any splits to unblock dynamic filter collection.`

**Review by arhimondr — APPROVED** [review]

**Comment by sopel39** [comment]
Thanks for review

**Pull request merged**
